File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size();
        int row_start = 0, row_end = n-1;
        int m = matrix[0].size();
       
        while(row_start<=row_end){
            if(target>=matrix[row_start][0] && target<=matrix[row_start][m-1]){
                int i=0, j = m-1;
                while(i<=j){
                    int mid = (i+j)/2;
                    if(matrix[row_start][mid]==target){
                        return true;
                    }
                    else if(matrix[row_start][mid]>target){
                        j=mid-1;
                    }
                    else i=mid+1;
                }
                return false;
            }
            if(target<matrix[row_start][0]) return false;
           while(row_start<= row_end && target>matrix[row_start][m-1])row_start++;
            
            
        }
    return false;
    }
};

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
//         base case.
        if(target< matrix[0][0]){
            return false;
        }
        int row_start = 0 , row_end = matrix.length-1;
        int row_ans=-1;
        
//          for row starting point..
        while(row_start<=row_end){
            int row_mid = (row_start+row_end)/2;
            if(matrix[row_mid][0]<= target){
                row_ans= row_mid;
                row_start = row_mid+1;
            }
            else row_end= row_mid-1;
        }
        System.out.print(row_ans);
//        
        int col_start=0, col_end = matrix[0].length-1;
        while(col_start<=col_end){
            int col_mid = (col_start+col_end)/2;
            if(matrix[row_ans][col_mid]==target){
                return true;
            }
            else if(matrix[row_ans][col_mid]> target){
                col_end = col_mid-1;
            }
            else col_start=col_mid+1;
        }
        return false;
        
    }
}